src/test/java/htsjdk/samtools/util/BlockCompressedInputStreamTest.java

```diff
@@ -86,4 +92,92 @@ public void available_should_return_number_of_bytes_left_in_current_block() thro
+        }
+        @Override
+        public int inflate(byte[] b, int off, int len) throws java.util.zip.DataFormatException {
+            inflateCalls++;
+            return super.inflate(b, off, len);
+        }
+        static int inflateCalls;
+    }
+
+    @DataProvider(name = "customInflaterInput")
+    public Object[][] customInflateInput() throws IOException {
+        final File tempFile = File.createTempFile("testCustomInflater.", ".bam");
+        tempFile.deleteOnExit();
+        System.out.println("Creating file " + tempFile);
+
+        final List<String> linesWritten = new ArrayList<>();
+        final BlockCompressedOutputStream bcos = new BlockCompressedOutputStream(tempFile, 5);
+        String s = "Hi, Mom!\n";
+        bcos.write(s.getBytes()); //Call 1
+        linesWritten.add(s);
+        s = "Hi, Dad!\n";
+        bcos.write(s.getBytes()); //Call 2
+        linesWritten.add(s);
+        bcos.flush();
+        final StringBuilder sb = new StringBuilder(BlockCompressedStreamConstants.DEFAULT_UNCOMPRESSED_BLOCK_SIZE * 2);
+        s = "1234567890123456789012345678901234567890123456789012345678901234567890\n";
+        while (sb.length() <= BlockCompressedStreamConstants.DEFAULT_UNCOMPRESSED_BLOCK_SIZE) {
+            sb.append(s);
+            linesWritten.add(s);
+        }
+        bcos.write(sb.toString().getBytes()); //Call 3
+        bcos.close();
+
+        final InflaterFactory myInflaterFactory = new InflaterFactory() {
+            public Inflater makeInflater(final boolean gzipCompatible) {
+                return new MyInflater(gzipCompatible);
+            }
+        };
+
+        // test catching null InflaterFactory
+        try {
+            BlockGunzipper.setDefaultInflaterFactory(null);
+            Assert.fail("Did not catch 'null' InflaterFactory");
+        }
+        catch (java.lang.IllegalArgumentException e) { /* expected */ }
+
+        BlockGunzipper.setDefaultInflaterFactory(myInflaterFactory);
+
+        return new Object[][]{
+                // use default InflaterFactory
+                {new BlockCompressedInputStream(new FileInputStream(tempFile), false), linesWritten, 4},
+                {new BlockCompressedInputStream(tempFile), linesWritten, 4},
+                {new AsyncBlockCompressedInputStream(tempFile), linesWritten, 4},
+                {new BlockCompressedInputStream(new URL("http://broadinstitute.github.io/picard/testdata/index_test.bam")), null, 21},
+                // provide InflaterFactory
+                {new BlockCompressedInputStream(new FileInputStream(tempFile), false, myInflaterFactory), linesWritten, 4},
+                {new BlockCompressedInputStream(tempFile, myInflaterFactory), linesWritten, 4},
+                {new AsyncBlockCompressedInputStream(tempFile, myInflaterFactory), linesWritten, 4},
+                {new BlockCompressedInputStream(new URL("http://broadinstitute.github.io/picard/testdata/index_test.bam"), myInflaterFactory), null, 21}
```

Comment:
@lbergelson suggests using a Supplier here to delay construction of the streams (and, therefore, any resulting exceptions) until the test case starts